airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/parquet/JsonSchemaType.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 Airbyte
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.destination.s3.parquet;
+
+import org.apache.avro.Schema;
+
+/**
+ * Mapping of JsonSchema types to Avro types.
+ */
+public enum JsonSchemaType {
+
+  STRING("string", Schema.Type.STRING),
+  NUMBER("number", Schema.Type.DOUBLE),
+  INTEGER("integer", Schema.Type.INT),
+  BOOLEAN("boolean", Schema.Type.BOOLEAN),
+  NULL("null", Schema.Type.NULL),
+  OBJECT("object", Schema.Type.RECORD),
```

Comment:
Probably just my unfamiliarity with Avro but what happens if there is an object with `additionalProperties: true` ? For example if we are replicating data from Mongo 😅 does that break serialization somehow? 

Also if the sub properties of the object are known, how come we are not encoding them into the RECORD defintiion? same question for arrays

Reply:
> what happens if there is an object with `additionalProperties: true`?

Currently those properties will be ignored. I have not looked into that yet. It seems like a 20% problem. But will do it.

> Also if the sub properties of the object are known, how come we are not encoding them into the RECORD defintion? same question for arrays

Known sub properties of an object is encoded into the record definition. Check the `nested_records` test case in `get_avro_schema.json`.

Array should also work. I will add a more comprehensive test case.

Reply:
Composite array support is done. Still need to work on `additionalProperties`.

Reply:
The additional properties can be stored in a map. However, all nested data will be stored as a serialized string, because there is no way we can tell the schema. In my opinion, Parquet is not suitable for schemaless stream, since those data will just be a json blob.

Reply:
The `additionalProperties` is tricky, because:
1. A schema is required for those additional properties, but there is no way to know that.
2. Even if we can create a schema for that, the current Json object to Avro record converter we are using does not support it (because of the first reason).

For example, given the following Json schema:

```json
{
  "type": "object",
  "properties": {
    "legalName": {
      "type": "string"
    }
  },
  "additionalProperties": {
    "type": "string"
  }
}
```

It means:
1. The `name` field is expected to be a `string`.
2. All additional properties should be `string`s.

This also means there is no way to have a Json schema for completely schemaless objects (e.g. those coming from mongo), as there is no single schema we can assign to the `additionalProperties`.

One solution I can think of is to move all fields not defined in the Json schema to one map as serialized string. In this way, we can have a defined Avro schema:

```json
{
  "type": "record",
  "fields": [
    { "name": "legalName", "type": "string" },
    { "name": "additionalProperties", "type": "map", "value": "string" }
  ]
}
```

Accordingly, the following Json object has one expected field (`legalName`) and two unexpected fields (`age` and `nickName`):

```json
{
  "legalName": "John Doe",
  "age": 46,
  "nickName": "John"
}
```

It can be converted to this Avro record:

```json
{
  "legalName": "John Doe",
  "additionalProperties": {
    "age": "46",
    "nickName": "John"
  }
}
```

Note that both `age` and `nickName` are moved to the `additionalProperties` field and their values become `string`.

Given the complexity, I'd like to roll out the MVP without supporting this keyword to give us more time to discuss it.